tests/unit/_modules/test_proposal.py

```diff
@@ -129,3 +146,142 @@ def test_create(self, output_helper):
         assert len(prop['nvme-ssd-spinner']) is 0
         assert len(prop['nvme-ssd']) is p.DEFAULT_DATA_R
         assert len(prop['nvme-spinner']) is p.DEFAULT_DATA_R
+
+
+    def replace_disk(self, type='hdd', ratio=2):
+        # idx 0-11 -> HDD
+        # idx 12-17 -> SSD
+        # idx 18 -> NVME
+        wal_db_replaced = None
+        hwinfo_out = OutputHelper().cephdisks_output
+        p = proposal.Proposal(OutputHelper().cephdisks_output, ratio=ratio)
+        old_prop = p.create()
+
+        # replace a HDD
+        hwinfo_out_new = hwinfo_out
+        replace_idx = random.randint(0, 11)
+        replaced_hdd = hwinfo_out_new[replace_idx]
+        del hwinfo_out_new[replace_idx]
+        new_disk = GenHwinfo(type=type).generate()
+
+        hwinfo_out_new.insert(random.randint(0, len(hwinfo_out)), new_disk)
+
+        # gen new proposal
+        new_p = proposal.Proposal(hwinfo_out_new)
+        new_prop = new_p.create()
+
+        for prop_name, prop in  old_prop.items():
+            for disk_set in old_prop[prop_name]:
+                # ensure that its a proposal with dedicated wal/db
+                if disk_set.keys()[0] in replaced_hdd['Device Files']:
+                    if prop_name in ['ssd-spinner', 'nvme-ssd-spinner', 'nvme-ssd', 'nvme-spinner']:
+                        wal_db_replaced = disk_set.values()[0]
+                        break
+
+        return old_prop, new_prop, replaced_hdd, wal_db_replaced
+
+    def is_deterministic(self, replaced_disk):
+        old_prop = replaced_disk[0]
+        new_prop = replaced_disk[1]
+        replaced_hdd = replaced_disk[2]
+        wal_db_replaced = replaced_disk[3]
+
+        for prop_name, prop in new_prop.items():
+            for disk_set in old_prop[prop_name]:
+                matching_disk_from_old_prop = [x for x in prop if disk_set == x]
+                osd_disk = disk_set.keys()[0]
+                wal_db = disk_set.values()[0]
+                # iterate over old proposal
+                # find disk in new proposal
+                # compare wal-db mapping
+
+                if not matching_disk_from_old_prop and osd_disk not in replaced_hdd['Device Files']:
+                    if 'GENERATED' in osd_disk:
+                        if wal_db != wal_db_replaced:
+                            return False, matching_disk_from_old_prop, osd_disk, wal_db, wal_db_replaced
+
+                if matching_disk_from_old_prop:
+                    assert osd_disk == matching_disk_from_old_prop[0].keys()[0]
+                    assert wal_db == matching_disk_from_old_prop[0].values()[0]
+                return True, [], osd_disk, wal_db, wal_db_replaced
```

Comment:
This return is too early, only one `disk_set in old_prop[prop_name]` gets checked.